File: deadchest-core/src/main/java/me/crylonz/deadchest/ChestData.java
package me.crylonz.deadchest;

import org.bukkit.Location;
import org.bukkit.configuration.serialization.SerializableAs;
import org.bukkit.entity.ArmorStand;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;

import java.util.*;

@SerializableAs("ChestData")
public final class ChestData {

    private List<ItemStack> inventory;
    private Location chestLocation;
    private String playerName;
    private String playerUUID;
    private Date chestDate;
    private boolean isInfinity;
    private boolean isRemovedBlock;
    private Location holographicTimer;
    private UUID holographicTimerId;
    private UUID holographicOwnerId;
    private String worldName;

    private int xpStored;

    public ChestData(ChestData chest) {
        this.inventory = chest.getInventory();
        this.chestLocation = chest.getChestLocation();
        this.playerName = chest.getPlayerName();
        this.playerUUID = chest.getPlayerUUID();
        this.chestDate = chest.getChestDate();
        this.isRemovedBlock = chest.isRemovedBlock();
        this.isInfinity = chest.isInfinity();
        this.holographicTimer = chest.getHolographicTimer();
        this.holographicTimerId = chest.getHolographicTimerId();
        this.holographicOwnerId = chest.getHolographicOwnerId();
        this.worldName = chest.getWorldName();
        this.xpStored = chest.getXpStored();
    }

    public ChestData(final Inventory inv,
                     final Location chestLocation,
                     final Player p,
                     final boolean isInfinity,
                     final ArmorStand asTimer,
                     final ArmorStand owner,
                     final int xpToStore) {

        if (p != null) {
            this.inventory = Arrays.asList(inv.getContents());
            this.chestLocation = chestLocation.clone();
            this.playerName = p.getName();
            this.playerUUID = String.valueOf(p.getUniqueId());
            this.chestDate = new Date();
            this.isInfinity = isInfinity;
            this.isRemovedBlock = false;
            this.holographicTimer = asTimer.getLocation().clone();
            this.holographicTimerId = asTimer.getUniqueId();
            this.holographicOwnerId = owner.getUniqueId();
            this.xpStored = xpToStore;
            if (chestLocation.getWorld() != null)
                this.worldName = chestLocation.getWorld().getName();
        }
    }

    public ChestData(final List<ItemStack> inventory,
                     final Location chestLocation,
                     final String playerName,
                     final String playerUUID,
                     final Date chestDate,
                     final boolean isInfinity,
                     final boolean isRemovedBlock,
                     final Location holographicTimer,
                     final UUID asTimerId,
                     final UUID asOwnerId,
                     final String worldName,
                     final int xpStored) {
        this.inventory = inventory;
        this.chestLocation = chestLocation;
        this.playerName = playerName;
        this.playerUUID = playerUUID;
        this.chestDate = chestDate;
        this.isRemovedBlock = isRemovedBlock;
        this.isInfinity = isInfinity;
        this.holographicTimer = holographicTimer;
        this.holographicTimerId = asTimerId;
        this.holographicOwnerId = asOwnerId;
        this.worldName = worldName;
        this.xpStored = xpStored;
    }

    public UUID getHolographicTimerId() {
        return holographicTimerId;
    }

    public UUID getHolographicOwnerId() {
        return holographicOwnerId;
    }

    public void setHolographicOwnerId(UUID holographicOwnerId) {
        this.holographicOwnerId = holographicOwnerId;
    }

    public void setHolographicTimerId(UUID holographicTimerId) {
        this.holographicTimerId = holographicTimerId;
    }

    public List<ItemStack> getInventory() {
        return inventory;
    }

    public void cleanInventory() {
        inventory = new ArrayList<>();
    }

    public void setInventory(List<ItemStack> inventory) {
        this.inventory = inventory;
    }

    public void setRemovedBlock(final boolean removedBlock) {
        this.isRemovedBlock = removedBlock;
    }

    public Location getChestLocation() {
        return chestLocation;
    }

    public String getPlayerName() {
        return playerName;
    }

    public String getPlayerUUID() {
        return playerUUID;
    }

    public Date getChestDate() {
        return chestDate;
    }

    public boolean isRemovedBlock() {
        return isRemovedBlock;
    }

    public boolean isInfinity() {
        return isInfinity;
    }

    public Location getHolographicTimer() {
        return holographicTimer;
    }

    public String getWorldName() {
        return worldName;
    }

    public boolean removeArmorStand() {
        final int radius = 1;
        final int armorStandShiftY = 1;

        if (chestLocation.getWorld() != null) {

            chestLocation.getChunk().setForceLoaded(true);

            Collection<Entity> entities = chestLocation.getWorld().getNearbyEntities(
                    new Location(
                            chestLocation.getWorld(),
                            chestLocation.getX(),
                            chestLocation.getY() + armorStandShiftY,
                            chestLocation.getZ()
                    ), radius, radius, radius);

            boolean isEmpty = entities.size() > 0;
            for (Entity entity : entities) {
                if (entity.getUniqueId().equals(holographicOwnerId) || entity.getUniqueId().equals(holographicTimerId)) {
                    entity.remove();
                }
            }

            if (isChunkForceLoaded()) {
                chestLocation.getWorld().unloadChunk(chestLocation.getBlockX() >> 4, chestLocation.getBlockZ() >> 4);
                chestLocation.getChunk().setForceLoaded(false);
            }

            return isEmpty;
        }
        return false;
    }

    public boolean isChunkLoaded() {
        return chestLocation.getWorld() == null ||
                chestLocation.getWorld().isChunkLoaded(
                        chestLocation.getBlockX() >> 4,
                        chestLocation.getBlockZ() >> 4
                );
    }

    public boolean isChunkForceLoaded() {
        return chestLocation.getWorld() == null ||
                chestLocation.getWorld().isChunkForceLoaded(
                        chestLocation.getBlockX() >> 4,
                        chestLocation.getBlockZ() >> 4
                );
    }

    public int getXpStored() {
        return xpStored;
    }

    public void setXpStored(int xpStored) {
        this.xpStored = xpStored;
    }

    enum Indexes {WORLD_NAME, LOC_X, LOC_Y, LOC_Z}
}

File: deadchest-core/src/main/kotlin/me/crylonz/deadchest/ChestData.kt
package me.crylonz.deadchest

import me.crylonz.deadchest.Indexes.*
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.configuration.serialization.ConfigurationSerializable
import org.bukkit.configuration.serialization.SerializableAs
import org.bukkit.entity.ArmorStand
import org.bukkit.entity.Player
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import java.util.*

@SerializableAs("ChestData")
data class ChestData(

    var inventory: List<ItemStack>,
    var chestLocation: Location,
    var playerName: String,
    var playerUUID: String,
    var chestDate: Date,
    var isInfinity: Boolean,
    var isRemovedBlock: Boolean,
    var holographicTimer: Location,
    var holographicTimerId: UUID,
    var holographicOwnerId: UUID,
    var worldName: String,
    var xpStored: Int,
) : ConfigurationSerializable {

    constructor(
        inv: Inventory,
        chestLocation: Location,
        player: Player,
        isInfinity: Boolean,
        armorStandTimer: ArmorStand,
        owner: ArmorStand,
        xpToStore: Int,
    ) : this(
        inventory = listOf<ItemStack>(*inv.contents),
        chestLocation = chestLocation.clone(),
        playerName = player.name,
        playerUUID = player.uniqueId.toString(),
        chestDate = Date(),
        isInfinity = isInfinity,
        isRemovedBlock = false,
        holographicTimer = armorStandTimer.location.clone(),
        holographicTimerId = armorStandTimer.uniqueId,
        holographicOwnerId = owner.uniqueId,
        worldName = player.world.name,
        xpStored = xpToStore
    )

    override fun serialize(): MutableMap<String, Any> {
        val map: MutableMap<String, Any> = HashMap()
        map["inventory"] = inventory
        map["chestLocation"] = "$worldName;${chestLocation.x};${chestLocation.y};${chestLocation.z}"
        map["playerName"] = playerName
        map["playerUUID"] = playerUUID
        map["chestDate"] = chestDate
        map["isRemovedBlock"] = isRemovedBlock
        map["isInfinity"] = isInfinity
        map["holographicTimer"] = "$worldName;${holographicTimer.x};${holographicTimer.y};${holographicTimer.z}"
        map["worldName"] = worldName
        map["as_timer_id"] = holographicTimerId.toString()
        map["as_owner_id"] = holographicOwnerId.toString()
        map["xpStored"] = xpStored
        return map
    }

    fun isChunkLoaded(): Boolean {
        return chestLocation.world == null ||
                chestLocation.world!!.isChunkLoaded(chestLocation.blockX shr 4, chestLocation.blockZ shr 4)
    }

    fun isChunkForceLoaded(): Boolean {
        return chestLocation.world == null ||
                chestLocation.world!!.isChunkForceLoaded(chestLocation.blockX shr 4, chestLocation.blockZ shr 4)
    }

    fun removeArmorStand(): Boolean {
        val radius = 1
        val armorStandShiftY = 1

        if (chestLocation.world != null) {
            chestLocation.chunk.isForceLoaded = true

            val entities = chestLocation.world!!.getNearbyEntities(
                Location(chestLocation.world, chestLocation.x, chestLocation.y + armorStandShiftY, chestLocation.z),
                radius.toDouble(), radius.toDouble(), radius.toDouble()
            )

            entities
                .filter { it.uniqueId == holographicOwnerId || it.uniqueId == holographicTimerId }
                .forEach { it.remove() }

            if (isChunkForceLoaded()) {
                chestLocation.world!!.unloadChunk(chestLocation.blockX shr 4, chestLocation.blockZ shr 4)
                chestLocation.chunk.isForceLoaded = false
            }

            return entities.isNotEmpty()
        }
        return false
    }

    fun cleanInventory() {
        inventory = emptyList()
    }

    companion object {
        @JvmStatic
        fun deserialize(map: Map<String, Any>): ChestData {
            val chestLocationSerialized = map["chestLocation"].toString().split(";")
            val holoLocationSerialized = map["holographicTimer"].toString().split(";")

            val chestLocation = Location(
                Bukkit.getWorld(chestLocationSerialized[WORLD_NAME.ordinal]),
                chestLocationSerialized[LOC_X.ordinal].toDouble(),
                chestLocationSerialized[LOC_Y.ordinal].toDouble(),
                chestLocationSerialized[LOC_Z.ordinal].toDouble()
            )

            val holoLocation = Location(
                Bukkit.getWorld(holoLocationSerialized[WORLD_NAME.ordinal]),
                holoLocationSerialized[LOC_X.ordinal].toDouble(),
                holoLocationSerialized[LOC_Y.ordinal].toDouble(),
                holoLocationSerialized[LOC_Z.ordinal].toDouble()
            )
            return ChestData(
                inventory = map["inventory"] as MutableList<ItemStack>,
                chestLocation = chestLocation,
                playerName = map["playerName"] as String,
                playerUUID = map["playerUUID"] as String,
                chestDate = map["chestDate"] as Date,
                isInfinity = map["isInfinity"] as Boolean,
                isRemovedBlock = map["isRemovedBlock"] != null && map["isRemovedBlock"] as Boolean, // compatibility under 4.14
                holographicTimer = holoLocation,
                holographicTimerId = UUID.fromString(map["as_timer_id"] as String),
                holographicOwnerId = UUID.fromString(map["as_owner_id"] as String),
                worldName = map["worldName"] as String,
                xpStored = if (map["xpStored"] != null) map["xpStored"] as Int else 0  // compatibility under 4.15
            )
        }
    }
}

internal enum class Indexes {
    WORLD_NAME,
    LOC_X,
    LOC_Y,
    LOC_Z
}
